### ISSUE TITLE
TRANSITION_EVENTS.JOIN ?

### ISSUE THREAD

**Issue by cyrilchapon** [uber/react-map-gl#567]
In here we'd like to achieve the following :

The user enters an adress :
- We perform an ajax query to reverse-geocode it (we get a point)
- We fly to that point with a transition
- **while** we do this, we perform another ajax query to get house boundaries of the user (we get a bbox)
- If the transition is still going, we "adjust" the ongoing transition to "finish" it with a sort fitBounds on the bbox
- If the transition is finished, we start a new one

Currently, I'm afraid the first use-case "adjusting" a transitition is not possible, am I right ?

I tried the following :

- Go for a transition (with the "set and forget" strategy)

```
    this.setState(prevState => ({
      viewport: {
        ...prevState.viewport,
        longitude: locationGeometry.coordinates[0],
        latitude: locationGeometry.coordinates[1],
        transitionDuration: MAP_LARGE_ANIMATION_DURATION,
        transitionInterpolator: new FlyToInterpolator(),
        transitionEasing: MAP_LARGE_ANIMATION_TRANSITION,
        transitionInterruption: TRANSITION_EVENTS.BREAK
      }
    }))
```

Then (1s later) change the viewport, without resetting transition props
```
    this.setState(prevState => {
      const parcelFitViewport = //Calculate viewport

      const {
        longitude,
        latitude,
        zoom
      } = parcelFitViewport

      return ({
        viewport: {
          ...prevState.viewport,
          longitude,
          latitude,
          zoom: zoom,
          // This is false in that use-case
          ...(prevState.initialZoomFinished ? {
            transitionDuration: MAP_SMALL_ANIMATION_DURATION,
            transitionInterpolator: new FlyToInterpolator(),
            transitionEasing: MAP_LARGE_ANIMATION_TRANSITION,
            transitionInterruption: TRANSITION_EVENTS.SNAP_TO_END
          } : {})
        }
      })
    })
```

But that doesn't work at all. The new viewport change **respects** the transition props, even if I don't re-set them. I also tried the 3 of the other transitionInteruptions strategies, but none of them accomplish what I want.

The troubles are :
- When I `setState(fitBoundsViewport)`, this entirely start a new transition (with the amount of time I set before), so if the first transition is 3000ms, with 2500 already elapsed, then the total transition time is roughly 5500ms if you follow me
- Therefore, I thought of "calculating the remaining time of the ongoing transition", and setting this time for the new one... but :
- There is no way to get the "remaining time of a transition" (the only events are "interrupted", "end" and "start", ie. no "tick"). Also, the onViewPortChange doesn't update the transition time. (it's 3000 then it's 0, without interpolated values)
- There is no way to "Join" the ongoing transition **ease**. If I "flyTo" with "easeInOut", the new transition **restarts** an easeInOut

An thoughts ?

**Comment by cyrilchapon** [comment]
Workaround :

- store the elapsed time in a variable, use `onTransitionStart` and an interval for it
- when you want to "join" a transition, hack it using that kind of function to create a custom ease for remainingTime :

```
const relativeEase = (remainingTime, baseTime, baseEase) => {
  return (time) => {
    // const timeRelativeToRemaining = remainingTime * time
    // const elapsedTime = baseTime - remainingTime
    // const timeRelativeToWholeTransition = elapsedTime + timeRelativeToRemaining
    // const relativeTime = timeRelativeToWholeTransition / baseTime

    const relativeTime = ((baseTime - remainingTime) + (remainingTime * time)) / baseTime

    return baseEase(relativeTime)
  }
}

this.setState({
  ...newViewport,
  transitionDuration: baseTransitionTime - elapsedTransitionTime,
  transitionInterpolator: new FlyToInterpolator(),
  transitionEasing: relativeEase(
    baseTransitionTime - elapsedTransitionTime,
    baseTransitionTime,
    baseEase
  )
})
```

**Comment by Pessimistress** [comment]
@cyrilchapon There might be complications that I haven't thought of, but looks to me that what you need is a new interruption handling mode?

```js
transitionInterruption: TRANSITION_EVENTS.UPDATE
```

**Comment by cyrilchapon** [comment]
Basically yes, I think reimplementing my workaround, and naming it `TRANSITION_EVENTS.JOIN` or `TRANSITION_EVENTS.UPDATE` would indeed be ideal.

I struggled mathematically to achieve this, though it's working like a charm with my solution, but it needs an extra `setInterval` to track the interaction timing (and join the previous with that timing).

To be clear, here I'm :
- Asking Is there a better known workaround ?
- Suggesting a new feature : `TRANSITION_EVENTS.JOIN`
- And since implementing my workaround lead me to separately, roughly, tracking the "remaining time", I'm also suggesting some sugar : a new event : `onTransitionTick` **OR** maybe the `onViewportChange` to directly update the `transitionDuration` progressively (and not brutaly from full duration to 0) to allow tracking of the concept of "remaining/elapsed duration of transition"

**Comment by Pessimistress** [comment]
@cyrilchapon This is landed and published in 4.0.0. Let me know if you need it cherry-picked back to 3.3.x.

**Comment by cyrilchapon** [comment]
Great 😀

I'll try and update to 4.0
